feedback/ngffeedback.cpp

```diff
@@ -0,0 +1,318 @@
+        ActiveEffect effect = m_activeEffects.take(id);
+        if (effect.customEffect)
+            m_customEffects.remove(effect.customEffect);
+    }
+}
+
+void NGFFeedback::completed(quint32 id)
+{
+    if (m_activeEffects.contains(id)) {
+        qCDebug(ngflc) << "Effect completed, id" << id;
+        ActiveEffect effect = m_activeEffects.take(id);
+        if (effect.customEffect)
+            m_customEffects.remove(effect.customEffect);
+    }
+}
+
+void NGFFeedback::playing(quint32 id)
+{
+    if (m_activeEffects.contains(id)) {
+        qCDebug(ngflc) << "Effect playing, id" << id;
+        m_activeEffects[id].state = QFeedbackEffect::Running;
+    }
+}
+
+void NGFFeedback::paused(quint32 id)
+{
+    if (m_activeEffects.contains(id)) {
+        qCDebug(ngflc) << "Effect paused, id" << id;
+        m_activeEffects[id].state = QFeedbackEffect::Paused;
+    }
+}
+
+bool NGFFeedback::play(QFeedbackEffect::Effect effect)
+{
+    quint32 id;
+    switch (effect) {
+    case QFeedbackEffect::Press:
+    case QFeedbackEffect::Release:
+    case QFeedbackEffect::PressWeak:
+    case QFeedbackEffect::ReleaseWeak:
+    case QFeedbackEffect::PressStrong:
+    case QFeedbackEffect::ReleaseStrong:
+    case QFeedbackEffect::DragStart:
+    case QFeedbackEffect::DragDropInZone:
+    case QFeedbackEffect::DragDropOutOfZone:
+    case QFeedbackEffect::DragCrossBoundary:
+        qCDebug(ngflc) << "Playing effect #" << effect << "(" << m_effects[effect] << ")";
+        id = m_client.play(m_effects[effect]);
+        if (!id)
+            qCWarning(ngflc) << "Could not play effect";
+        else
+            m_activeEffects.insert(id, { QFeedbackEffect::Running, nullptr });
+        return true;
+    case QFeedbackEffect::Appear:
+    case QFeedbackEffect::Disappear:
+    case QFeedbackEffect::Move:
+        qCDebug(ngflc) << "Unsupported effect #" << effect;
+        break;
+    default:
+        qCDebug(ngflc) << "Unknown or undefined effect #" << effect;
+        break;
+    }
+
+    return false;
+}
+
+QFeedbackInterface::PluginPriority NGFFeedback::pluginPriority()
+{
+    return QFeedbackInterface::PluginLowPriority;
+}
+
+QList<QFeedbackActuator*> NGFFeedback::actuators()
+{
+    return QList<QFeedbackActuator*>() << m_actuator;
+}
+
+void NGFFeedback::setActuatorProperty(const QFeedbackActuator &, ActuatorProperty prop, const QVariant &value)
+{
+    if (prop == ActuatorProperty::Enabled) {
+        bool old = m_actuatorEnabled;
+        m_actuatorEnabled = value.toBool();
+        if (old != m_actuatorEnabled && !m_actuatorEnabled) {
+            /* Stop all effects */
+            for (auto it = m_activeEffects.begin(); it != m_activeEffects.end(); ++it) { // TODO: Would it make sense to iterate backwards?
+                quint32 id = it.key();
+                ActiveEffect effect = it.value();
+                it = m_activeEffects.erase(it); // Remove it already to avoid changing m_activeEffects
```

Comment:
And need to fix this. I was thinking about it and then forgot to make sure it works correctly. The problem is that erase() returns iterator to next item, not previous so doing ++it will skip one.

Reply:
Yeah, well, ended up rewriting a good part of the plugin so maybe it works now.